exporter/otlpexporter/exporter.go

```diff
@@ -0,0 +1,274 @@
+// Copyright  OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package otlpexporter
+
+import (
+	"context"
+	"errors"
+	"sync"
+	"time"
+	"unsafe"
+
+	"github.com/cenkalti/backoff"
+	otlpmetriccol "github.com/open-telemetry/opentelemetry-proto/gen/go/collector/metrics/v1"
+	otlptracecol "github.com/open-telemetry/opentelemetry-proto/gen/go/collector/trace/v1"
+	"google.golang.org/genproto/googleapis/rpc/errdetails"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+
+	"github.com/open-telemetry/opentelemetry-collector/config/configgrpc"
+)
+
+type exporterImp struct {
+	// Input configuration.
+	config *Config
+
+	// Prepared dial options.
+	dialOpts []grpc.DialOption
+
+	startOnce sync.Once
+	stopOnce  sync.Once
+
+	started bool
+	stopped bool
+
+	// mutex to protects bool flags.
+	mutex sync.RWMutex
+
+	stopCh                     chan bool
+	disconnectedCh             chan bool
+	backgroundConnectionDoneCh chan bool
+
+	// gRPC clients and connection.
+	traceExporter     otlptracecol.TraceServiceClient
+	metricExporter    otlpmetriccol.MetricsServiceClient
+	grpcClientConn    *grpc.ClientConn
+	lastConnectErrPtr unsafe.Pointer
+}
+
+var (
+	errAlreadyStarted = errors.New("already started")
+	errNotStarted     = errors.New("not started")
+	errTimeout        = errors.New("Timeout")
+	errStopped        = errors.New("Stopped")
+	errFatalError     = errors.New("Fatal error sending to server")
+)
+
+// Crete new exporter and start it. The exporter will begin connecting but
+// this function may return before the connection is established.
+func newExporter(config *Config) (*exporterImp, error) {
+	e := &exporterImp{}
+	e.config = config
+	e.disconnectedCh = make(chan bool, 1)
+	e.stopCh = make(chan bool)
+	e.backgroundConnectionDoneCh = make(chan bool)
+
+	var err error
+	e.dialOpts, err = configgrpc.GrpcSettingsToDialOptions(e.config.GRPCSettings)
+	if err != nil {
+		return nil, err
+	}
+
+	if err := e.start(); err != nil {
```

Comment:
TODO: since we have a start method for exporters we should make this call there and not start a goroutine at instance creation.